compiler/identify_token.cpp

```diff
@@ -27,120 +28,140 @@ void init_isIdChar() {
   isIdChar['_'] = 1;
 }
 
+/*
+    The function intializes the isBlankChar array. If a character c:
+    ---> is newline, space or tab character, then isBlankChar[c]=1
+    ---> Otherwise, isBlankChar[c]=0
+*/
+void init_isBlankChar() {
+  isBlankChar[' '] = 1;
+  isBlankChar['\t'] = 1;
+  isBlankChar['\n'] = 1;
+}
+
 /*
 The function sets the kind member of the global variable token to the fist token
 found in the input string, if that token is a keyword or a special character.
 If a blank character, an integer literal, an identifier or a non-DIL string is
 encoutered, kind is set to UNDEFINED.
 */
 void identify_token(const char *input) {
-  switch (input[0]) {
+  int ind = 0;
+  while (isBlankChar[input[ind]])
+    ind++;
+  switch (input[ind]) {
```

Comment:
You really strain the compiler for no reason :P Instead of doing `ind++`, do `input++` and the code stays the same as before.

Reply:
This is because `input[ind]` is equivalent to `*(input+ind)` right? So I could have `isBlankChar[*input]` instead (1 addition less). One question here is what happens if I have `input[0]` instead of `*input`. Is the addition operation of `input+0` calculated?

Reply:
> This is because input[ind] is equivalent to *(input+ind) right? So I could have isBlankChar[*input] instead (1 addition less).

Yes, but consider that the code now has way more additions. While, if you just advance `input` instead of indexing it, the code is essentially the same as before.
Note that the compiler will probably do it for you, but why make its life harder and the readability of the code worse?

> what happens if I have input[0] instead of *input. Is the addition operation of input+0 calculated?

No, it's basically the same thing. Theoretically it should, but the compiler doesn't do it obviously.
I think that in such cases, if you're interested, you can just check the assembly (or IR if you're using LLVM). And ask me if you don't understand something. :)